atlp-rwanda/stackup2-barefoot-backend
# #171380750 User logout functionality

#### What does this PR do?
Implement logout functionality
#### Description of Task to be completed?
* Implement endpoint unit tests
* Create logout controller
* Create helper functions, middlewares, status codes, and custom messages.
* Install required libraries
* Create logout route
* Implement JsDoc and endpoint documentation
* Create Cron job to remove expired tokens from redis database everyday at midnight.
#### How should this be manually tested?
* Set `REDIS_PORT` variable in `.env` to `REDIS_PORT=6379`
1. Open your terminal and run `git clone https://github.com/Stackup-Rwanda/stackup2-barefoot-backend/`
2. run `cd stackup2-barefoot-backend` to go into the root directory of the project
3. run `git checkout -b ft-user-logout-171380750 origin/ft-user-logout-171380750` to pull this branch and switch to it
4. run `npm install` to install the dependencies
5. Make sure you have signed up then using POSTMAN, make a request at `http://localhost:3000/api/auth/login` to login
6. Copy the token in the response body
7. Make a GET request at `http://localhost:3000/api/auth/logout` to log out and make sure you include the token copied above in the following header: 
`Authorization: Bearer token`
8. Hit send and you should get a response of `Logout successful`
9. Hit send again and you should get a response of `Please login first`
* To test the cron job, follow these steps:
1. Set your env variabe `SIGN_UP_TOKEN_EXPIRES_IN` to `2m`
2. Go to `src > taskScheduler.js` and change this pattern `'* 59 23 * * *'` to `'*/5 * * * *'` to make this job run every 5 minutes
3. Run the server with `npm run dev` and follow the steps above from 5. to 8 to logout
4. Open your terminal and run `redis-cli`, you should now be inside the redis CLI.
5. Type `smembers token` then press Enter and you should see your token listed there
6. Wait for 5 minutes and type `smembers token` then press Enter again and your token should have been deleted.
#### Any background context you want to provide?
* You need to install [Node.js](https://nodejs.org/en/) and [git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) before testing.
* You need to have **PostgeSQL** database installed and configured on your computer
* You need to have **Redis** server installed on your system. Click [here](https://redislabs.com/blog/redis-on-windows-10/) for steps on how to install Redis
#### What are the relevant pivotal tracker stories?
[#171380750](https://www.pivotaltracker.com/story/show/171380750)
#### Screenshots (if appropriate)
![Screenshot from 2020-03-18 11-07-25](https://user-images.githubusercontent.com/45935765/76944321-7c583880-6909-11ea-9546-b0d2a6505848.png)
![Screenshot from 2020-03-18 11-07-38](https://user-images.githubusercontent.com/45935765/76944333-7febbf80-6909-11ea-9b8e-151548d6c6c4.png)
![Screenshot from 2020-03-18 11-10-21](https://user-images.githubusercontent.com/45935765/76944336-824e1980-6909-11ea-94b6-2b3ad0ef5927.png)

#### Questions:
N/A
